storage/src/storage/sqlite/sync.rs

```diff
@@ -0,0 +1,923 @@
+        INNER JOIN blocks ON blocks.id = transaction_blocks.id
+        WHERE blocks.canon_height IS NOT NULL
+        ORDER BY blocks.canon_height ASC, transaction_blocks.block_order ASC
+        ",
+        )?;
+        let digests = stmt
+            .query_map([], |row| Ok([row.get(0)?, row.get(1)?]))?
+            .collect::<rusqlite::Result<Vec<[Digest; 2]>>>()?
+            .into_iter()
+            .flatten()
+            .collect();
+        Ok(digests)
+    }
+
+    fn get_serial_numbers(&mut self) -> Result<Vec<Digest>> {
+        let mut stmt = self.conn.prepare_cached(
+            r"
+        SELECT
+        transactions.old_serial_number1,
+        transactions.old_serial_number2
+        FROM transactions
+        INNER JOIN transaction_blocks ON transaction_blocks.transaction_id = transactions.id
+        INNER JOIN blocks ON blocks.id = transaction_blocks.id
+        WHERE blocks.canon_height IS NOT NULL
+        ORDER BY blocks.canon_height ASC, transaction_blocks.block_order ASC
+        ",
+        )?;
+        let digests = stmt
+            .query_map([], |row| Ok([row.get(0)?, row.get(1)?]))?
+            .collect::<rusqlite::Result<Vec<[Digest; 2]>>>()?
+            .into_iter()
+            .flatten()
+            .collect();
+        Ok(digests)
+    }
+
+    fn get_memos(&mut self) -> Result<Vec<Digest>> {
+        let mut stmt = self.conn.prepare_cached(
+            r"
+        SELECT
+        transactions.memo
+        FROM transactions
+        INNER JOIN transaction_blocks ON transaction_blocks.transaction_id = transactions.id
+        INNER JOIN blocks ON blocks.id = transaction_blocks.id
+        WHERE blocks.canon_height IS NOT NULL
+        ORDER BY blocks.canon_height ASC, transaction_blocks.block_order ASC
+        ",
+        )?;
+        let digests = stmt
+            .query_map([], |row| row.get(0))?
+            .collect::<rusqlite::Result<Vec<Digest>>>()?;
+        Ok(digests)
+    }
+
+    fn get_ledger_digests(&mut self) -> Result<Vec<Digest>> {
+        let mut stmt = self.conn.prepare_cached(
+            r"
+        SELECT
+        blocks.canon_ledger_digest
+        FROM blocks
+        WHERE blocks.canon_height IS NOT NULL
+        ORDER BY blocks.canon_height ASC
+        ",
+        )?;
+        let digests = stmt
+            .query_map([], |row| row.get(0))?
+            .collect::<rusqlite::Result<Vec<Digest>>>()?;
+        Ok(digests)
+    }
+
+    fn reset_ledger(
+        &mut self,
+        _commitments: Vec<Digest>,
+        _serial_numbers: Vec<Digest>,
+        _memos: Vec<Digest>,
+        _digests: Vec<Digest>,
+    ) -> Result<()> {
+        unimplemented!()
+    }
+
+    fn get_canon_blocks(&mut self, limit: Option<u32>) -> Result<Vec<SerialBlock>> {
+        let digests = self.get_block_hashes(limit, BlockFilter::CanonOnly(BlockOrder::Unordered))?;
+        // this is intentionally N+1 query since this is not a critical performance function and its easy
```

Comment:
do we intend to have a `WHERE IN` query later on, just to squeeze out maximum perf? it could be left as a `TODO` or posted as a `help-wanted` GitHub issue